# Déjame escuchar música

Los datos están almacenados en el archivo `/datasets/music_project_en.csv`.

¡Ahora sí, manos al código!


## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Etapa 1.1. Necesitarás `pandas`, así que impórtalo.

In [1]:
# Importa pandas
import pandas as pd

Etapa 1.2. Lee el archivo `music_project_en.csv` de la carpeta `datasets/` y guárdalo en la variable `df`:

In [ ]:
# Lee el archivo y almacénalo en df
df = pd.read_csv('datasets/music_project_en.csv')

Etapa 1.3. Muestra las 10 primeras filas de la tabla:

In [3]:
# Obtén las 10 primeras filas de la tabla df
print(df.head(10))

     userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5

Etapa 1.4. Obtén la información general sobre la tabla con el método info().

In [4]:
# Obtén la información general sobre nuestros datos
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype
---  ------    --------------  -----
 0     userID  65079 non-null  str  
 1   Track     63736 non-null  str  
 2   artist    57512 non-null  str  
 3   genre     63881 non-null  str  
 4     City    65079 non-null  str  
 5   time      65079 non-null  str  
 6   Day       65079 non-null  str  
dtypes: str(7)
memory usage: 7.5 MB


Estas son nuestras observaciones sobre la tabla. Contiene siete columnas que almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la canción;
- `'City'`: ciudad del usuario;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver dos problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. `Identifica tú mismo el segundo problema y escríbelo aquí.`
 - Algunos nombres de columnas tienen espacios innecesarios al inicio o al final.
 - Hay inconsistencia de estilo: userID, Track, City, Day, artist, genre, time. Lo ideal sería usar un formato uniforme, por ejemplo todo en minúsculas y sin espacios.

### Escribe algunas observaciones por tu parte. Contesta a las siguientes preguntas: <a id='data_review_conclusions'></a>

`1.   ¿Qué tipo de datos hay en las filas? ¿Cómo podemos saber qué almacenan las columnas?`

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`

Escribe aquí tus respuestas:

1. Cada fila representa una reproducción de una canción por un usuario. Las columnas almacenan información sobre el usuario, la canción, el artista, el género, la ciudad, la hora y el día de reproducción. Podemos saber qué almacena cada columna revisando la información devuelta por info(), en este caso vemos que todas las columnas son tipo 'object'.

2. Sí parece haber suficientes datos para analizar patrones de reproducción por ciudad, día, hora, artista o género. Sin embargo, seguramente, es necesario limpiar los datos y revisar los valores ausentes.

3. Sí. Hay valores ausentes en las columnas Track, artist y genre. También hay problemas en los nombres de las columnas: algunas tienen mayúsculas, otras minúsculas y algunas parecen tener espacios innecesarios. Por ahora no sabemos si hay duplicados; habría que comprobarlo con df.duplicated().sum().

## Etapa 2. Preprocesamiento de los datos <a id='data_preprocessing'></a>

Tu objetivo aquí es preparar los datos para analizarlos.
El primer paso es resolver los problemas con los encabezados. Después podemos avanzar a los valores ausentes y duplicados. ¡Empecemos!

Vamos a corregir el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Etapa 2.1. Muestra los encabezados de la tabla (los nombres de las columnas):

In [5]:
# Muestra los nombres de las columnas
print(df.columns)

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='str')


Vamos cambiar los encabezados de la tabla siguiendo las reglas estilísticas convencionales:
*   Todos los caracteres deben ser minúsculas.
*   Elimina los espacios.
*   Si el nombre tiene varias palabras, utiliza snake_case, es decir, añade un guion bajo ( _ ) entre las palabras en lugar de un espacio.



Etapa 2.2. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [6]:
# Bucle que itera sobre los encabezados y los pone todos en minúsculas

col_names_lower = []

for old_name in df.columns:
    name_lower = old_name.lower()
    col_names_lower.append(name_lower)

df.columns = col_names_lower

print(df.columns)

Index(['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day'], dtype='str')


Etapa 2.3. Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas y muestra los nombres de las columnas de nuevo:

In [7]:
# Bucle que itera sobre los encabezados y elimina los espacios

col_names_stripped = []

for old_name in df.columns:
    name_stripped = old_name.strip()
    col_names_stripped.append(name_stripped)

df.columns = col_names_stripped

print(df.columns)

Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='str')


Etapa 2.4. Necesitamos aplicar la regla de snake_case en la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [8]:
# Cambia el nombre de la columna "userid"
df.rename(columns={'userid': 'user_id'}, inplace=True)

print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='str')


Etapa 2.5. Comprueba el resultado. Muestra los encabezados una vez más:

In [9]:
# Comprueba el resultado: lista de encabezados
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='str')


### Valores ausentes <a id='missing_values'></a>
 Etapa 2.5. Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos para obtener el número de valores ausentes.

In [10]:
# Calcula el número de valores ausentes
print(df.isna().sum())
print(df.isnull().sum())

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64
user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


Etapa 2.6. Sustituye los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`.

1. Crea una lista llamada columns_to_replace que contenga los nombres de las columnas 'track', 'artist' y 'genre'.

2. Usa un bucle for para iterar sobre cada columna en columns_to_replace.

3. Dentro del bucle, sustituye los valores ausentes en cada columna con el string `'unknown'`.

In [11]:
# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'

columns_to_replace = ['track', 'artist', 'genre']

for column in columns_to_replace:
    df[column] = df[column].fillna('unknown')

Etapa 2.7. Ahora comprueba el resultado para asegurarte de que no falten valores ausentes por reemplazar en el conjunto de datos. Para ello, cuenta los valores ausentes una vez más.

In [12]:
# Cuenta los valores ausentes
print(df.isna().sum())

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


### Duplicados <a id='duplicates'></a>
Etapa 2.8. Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos para obtener la cantidad de duplicados explícitos.

In [13]:
# Cuenta los duplicados explícitos
print(df.duplicated().sum())
print(len(df[df.duplicated()]))

3826
3826


Etapa 2.9. Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [14]:
# Elimina los duplicados explícitos
df = df.drop_duplicates().reset_index(drop=True)

Etapa 2.10. Comprobemos ahora si conseguimos eliminar todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [15]:
# Comprueba de nuevo si hay duplicados
print(df.duplicated().sum())

0


Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Etapa 2.11. Primero debemos mostrar una lista de nombres de géneros únicos, por orden alfabético. Para ello:
1. Extrae la columna `genre` del DataFrame.
2. Llama al método que devolverá todos los valores únicos en la columna extraída.


In [16]:
# Inspecciona los nombres de géneros únicos
print(df['genre'].sort_values().unique())

<ArrowStringArray>
[       'acid',    'acoustic',      'action',       'adult',      'africa',
   'afrikaans', 'alternative',     'ambient',   'americana',    'animated',
 ...
       'urban',       'uzbek',     'variété',          'vi',   'videogame',
       'vocal',     'western',       'world',   'worldbeat',         'ïîï']
Length: 269, dtype: str


Etapa 2.12. Vamos a examinar la lista para identificar **duplicados implícitos** del género `hiphop`, es decir, nombres mal escritos o variantes que hacen referencia al mismo género musical.

Los duplicados que encontrarás son:

* `hip`  
* `hop`  
* `hip-hop`  

Para solucionarlo, vamos a crear una función llamada `replace_wrong_values()`.


1. Define una función llamada `replace_wrong_values()` que reciba los siguientes parámetros:

* `df`: el DataFrame a modificar
* `column`: el nombre de la columna a trabajar
* `wrong_values`: una lista con los valores incorrectos
* `correct_value`: el valor correcto para reemplazar

2. Dentro de la función, usa un bucle `for` para iterar sobre cada valor incorrecto y aplicar `.replace()`.


In [17]:
# Función para reemplazar los duplicados implícitos
def replace_wrong_values(df, column, wrong_values, correct_value):
    for wrong_value in wrong_values:
        df[column] = df[column].replace(wrong_value, correct_value)
    return df

Etapa 2.13. Ahora, llama a la función pasando:

* `df` como el DataFrame
* `'genre'` como nombre de columna
* `['hip', 'hop', 'hip-hop']` como lista de valores incorrectos
* `'hiphop'` como valor correcto


In [18]:
# Elimina los duplicados implícitos
replace_wrong_values(df, 'genre', ['hip', 'hop', 'hip-hop'], 'hiphop')

,user_id,track,artist,genre,city,time,day
0,FFB692EC,Kamigata To Boots,The Mass Missile,rock,Shelbyville,20:28:33,Wednesday
1,55204538,Delayed Because of Accident,Andreas Rönnberg,rock,Springfield,14:07:09,Friday
2,20EC38,Funiculì funiculà,Mario Lanza,pop,Shelbyville,20:58:07,Wednesday
3,A3DD03C9,Dragons in the Sunset,Fire + Ice,folk,Shelbyville,08:37:09,Monday
4,E2DC1FAE,Soul People,Space Echo,dance,Springfield,08:34:34,Monday
...,...,...,...,...,...,...,...
61248,729CBB09,My Name,McLean,rnb,Springfield,13:32:28,Wednesday
61249,D08D4A55,Maybe One Day (feat. Black Spade),Blu & Exile,hiphop,Shelbyville,10:00:00,Monday
61250,C5E3A0D5,Jalopiina,unknown,industrial,Springfield,20:09:26,Friday
61251,321D0506,Freight Train,Chas McDevitt,rock,Springfield,21:43:59,Friday


Etapa 2.14. Asegúrate de que los nombres duplicados se hayan eliminado. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [19]:
# Comprueba de nuevo los duplicados implícitos
print(df['genre'].sort_values().unique())

<ArrowStringArray>
[       'acid',    'acoustic',      'action',       'adult',      'africa',
   'afrikaans', 'alternative',     'ambient',   'americana',    'animated',
 ...
       'urban',       'uzbek',     'variété',          'vi',   'videogame',
       'vocal',     'western',       'world',   'worldbeat',         'ïîï']
Length: 266, dtype: str


## Etapa 3. Análisis

### Tarea: Comparar el comportamiento de los usuarios en las dos ciudades <a id='activity'></a>

Queremos analizar si hay diferencias en la cantidad de canciones reproducidas en Springfield y Shelbyville. Para ello, usaremos los datos de dos días de la semana: lunes y viernes.

Compararemos cuántas canciones se escucharon en cada ciudad durante esos días para identificar posibles patrones de comportamiento.

Sigue estos tres pasos para organizar tu análisis:

- Dividir: agrupa los datos por ciudad.

- Aplicar: cuenta cuántas canciones se reproducen en cada grupo.

- Combinar: presenta los resultados de forma que se puedan comparar fácilmente ambas ciudades.

Repite este proceso por separado para cada uno de los dos días.

Etapa 3.1.
Cuenta cuántas canciones se reprodujeron en cada ciudad utilizando la columna `'track'` como referencia.

In [20]:
# Cuenta las canciones reproducidas en cada ciudad
print(df.groupby('city')['track'].count())
print(df[df['day'] == 'Monday'].groupby('city')['track'].count())
print(df[df['day'] == 'Friday'].groupby('city')['track'].count())

city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64
city
Shelbyville     5614
Springfield    15740
Name: track, dtype: int64
city
Shelbyville     5895
Springfield    15945
Name: track, dtype: int64


Etapa 3.2. Redacta brevemente tus observaciones sobre los resultados.

¿Qué diferencias encontraste entre Springfield y Shelbyville? ¿A qué podrían deberse?

Escribe tus observaciones aquí.

La diferencia es bastante grande: Springfield tiene casi el triple de reproducciones en ambos días.
También se puede inferir que esos dos días representan el 62% para Shelbyville y 74% para Springfield del total de reproducciones.
Podría haber más numero de usuarios registrados en Spriingfield que en Shelbyville



Etapa 3.3.
Agrupa los datos por día de la semana y cuenta cuántas canciones se reprodujeron los lunes y viernes.



In [21]:
# Calcula las canciones reproducidas en cada uno de los dos días
print(df[df['day'].isin(['Monday', 'Friday'])].groupby('day')['track'].count())
print(df[df['day'].isin(['Monday', 'Friday'])]['track'].count())

day
Friday    21840
Monday    21354
Name: track, dtype: int64
43194


Etapa 3.4. Describe brevemente qué observaste al comparar los lunes y viernes.

¿Hubo un día con más actividad? ¿Cambia algo si analizas cada ciudad por separado?

Escribe tus observaciones aquí.

Entre ambos días es prácticamente la misma cantidad. Como no existe diferencia sustancial entre ambos días, el análisis entre ciudades nos indica una tendencia mayor.

Etapa 3.5

Ahora vamos a combinar dos criterios: día y ciudad.

Crea una función llamada `number_tracks()` que reciba dos parámetros:

* `day`: un día de la semana (por ejemplo, `'Monday'`)
* `city`: el nombre de una ciudad (por ejemplo, `'Springfield'`)

Dentro de la función:

1. Filtra el DataFrame por el día.
2. Luego, filtra por la ciudad.
3. Cuenta cuántas veces aparece `'user_id'` en ese filtro.
4. Devuelve ese número como resultado.


In [22]:
# Declara la función number_tracks() con dos parámetros: day= y city=.

def number_tracks(day, city):
    df_by_day = df[df['day'] == day]
    df_by_city = df_by_day[df_by_day['city'] == city]
    track_count = df_by_city['user_id'].count()
    return track_count

Etapa 3.6. Llama a `number_tracks()` cuatro veces: una por ciudad en cada uno de los dos días.

In [23]:

# El número de canciones reproducidas en Springfield el lunes
print(number_tracks('Monday', 'Springfield'))


15740


In [24]:

# El número de canciones reproducidas en Shelbyville el lunes
print(number_tracks('Monday', 'Shelbyville'))


5614


In [25]:

# El número de canciones reproducidas en Springfield el viernes
print(number_tracks('Friday', 'Springfield'))



15945


In [26]:
# El número de canciones reproducidas en Shelbyville el viernes
print(number_tracks('Friday', 'Shelbyville'))

5895


# Conclusiones <a id='end'></a>

## Escribe tus conclusiones finales sobre el análisis

Redacta un resumen breve y claro de los hallazgos obtenidos durante el proceso de análisis. Tu conclusión debe:

* Mencionar los principales patrones que observaste en los datos.
* Identificar cualquier problema encontrado y cómo lo solucionaste.
* Explicar cómo estas acciones ayudaron a mejorar la calidad del análisis.

Reflexiona también sobre la pregunta central:

> *¿Los datos muestran que el comportamiento de los usuarios —en cuanto a la música que escuchan— varía según la ciudad y el día de la semana?*

Apoya tu respuesta con ejemplos concretos de los resultados obtenidos.


Detalla aquí tus conclusiones.

Durante el análisis observé que existen diferencias claras en la cantidad de reproducciones entre Springfield y Shelbyville. En ambos días analizados, Springfield tuvo muchas más reproducciones que Shelbyville. Por ejemplo, el lunes Springfield registró 15,740 reproducciones, mientras que Shelbyville tuvo 5,614. El viernes ocurrió algo similar: Springfield tuvo 15,945 reproducciones y Shelbyville 5,895. Esto sugiere que los usuarios de Springfield muestran una actividad mucho mayor en la plataforma.
También encontré varios problemas en los datos. Primero, los nombres de las columnas tenían estilos inconsistentes: algunas estaban en mayúsculas, otras en minúsculas y algunas tenían espacios innecesarios. Para solucionarlo, convertí los nombres a minúsculas, eliminé espacios y cambié userid a user_id. Después, detecté valores ausentes en las columnas track, artist y genre, y los reemplacé con 'unknown'. Además, eliminé duplicados explícitos y corregí duplicados implícitos en la columna genre, reemplazando valores como 'hip', 'hop' y 'hip-hop' por 'hiphop'.
Estas acciones mejoraron la calidad del análisis porque permitieron trabajar con columnas uniformes, evitar errores al agrupar datos y reducir inconsistencias que podían distorsionar los resultados. Por ejemplo, si no se corregían los géneros duplicados, el mismo género musical podía aparecer como categorías distintas.
En conclusión, los datos sí muestran que el comportamiento de los usuarios varía según la ciudad y el día de la semana. Springfield presenta una mayor cantidad de reproducciones que Shelbyville tanto el lunes como el viernes. Sin embargo, la diferencia entre lunes y viernes dentro de cada ciudad no es tan grande: en Springfield las reproducciones fueron muy parecidas entre ambos días, y lo mismo ocurrió en Shelbyville. Por lo tanto, la ciudad parece influir más que el día de la semana en la cantidad de música escuchada.